Add StringTensor

joey12300 opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
New features

### PR changes
<!-- One of [ OPs | APIs | Docs | Others ] -->
OPs

### Describe
<!-- Describe what this PR does -->
该PR仅修改C++层的代码，暂时不涉及Python API的修改

1. 新增StringTensor C++层数据结构
2. 新增一个字符串Kernel——大小写转换，支持CPU，GPU上运行
3. 新增字符串C++ API的生成脚本，可以通过配置yaml生成字符串C++ API

TODO：
1. 新增字符串动态图API，Python-C接口生成
2. 补充StringTensor的Pybind接口，用于在Python层创建StringTensor。
3. 补充更多字符串算子

joey12300 committed:
add string tensor and case convert kernels

joey12300 committed:
Add strings empty kernel; Reorganize the structure of case convert kernel

joey12300 committed:
Add string infermeta

joey12300 committed:
Update mutable_data of string tensor

joey12300 committed:
merge develop

joey12300 committed:
rename kernel name

joey12300 committed:
add string copy tmp

joey12300 committed:
Fix strings copy device bug

joey12300 committed:
add utf8 gpu converter

joey12300 committed:
add string tensor c++ api

joey12300 committed:
Merge remote-tracking branch 'upstream' into add_string_tensor

joey12300 committed:
Remove mutable_data of string tensor

joey12300 committed:
update string tensor interface

joey12300 committed:
remove charcases_flag.h

joey12300 committed:
Merge branch 'develop' into add_string_tensor

Conflicts:
	cmake/phi.cmake
	paddle/phi/CMakeLists.txt
	paddle/phi/api/CMakeLists.txt
	paddle/phi/api/lib/CMakeLists.txt
	paddle/phi/api/lib/utils/CMakeLists.txt
	paddle/phi/common/data_type.h
	paddle/phi/core/kernel_utils.h
	paddle/phi/tests/api/CMakeLists.txt
	paddle/phi/tests/kernels/CMakeLists.txt

joey12300 committed:
remove some fluid headers

joey12300 committed:
Add make_ddim

joey12300 committed:
__HIPCC__ -> PADDLE_WITH_HIP

joey12300 committed:
remove fluid headers

joey12300 committed:
fix cpu compile

joey12300 committed:
remove std::hash

joey12300 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_string_tensor

joey12300 committed:
fix cmakelist conflics

joey12300 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_string_tensor

joey12300 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_string_tensor

chenwhql commented on paddle/phi/common/data_type.h:
```diff
@@ -45,6 +47,7 @@ enum class DataType {
   FLOAT64,
   COMPLEX64,
   COMPLEX128,
+  STRING,
```
建议在上方多写一些注释，说明
1. 为什么这里string需要作为一种dtype？
2. string是一种什么样的dtype？

joey12300 replied:
已添加注释，并将STRING改为PSTRING

joey12300 committed:
Fix cudaMalloc

joey12300 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_string_tensor

Conflicts:
	paddle/phi/tests/api/CMakeLists.txt

joey12300 committed:
Remove strings/impl directory

joey12300 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_string_tensor

joey12300 committed:
Fix infrt/get_phi_kernel_info.py;Add custom_kernels deps

joey12300 committed:
Add empty kernel test

Shixiaowei02 commented on paddle/phi/common/pstring.h:
```diff
@@ -0,0 +1,475 @@
+  HOSTDEVICE bool operator!=(const pstring& o) const;
+
+  // Conversion Operators
+  // TODO(b/147740521): Make explicit.
```
注释需要清理

joey12300 replied:
Done

joey12300 committed:
Remove some comments

joey12300 committed:
Modify lower/upper api encoding type: string->bool

chenwhql commented on paddle/phi/common/data_type.h:
```diff
@@ -82,22 +87,23 @@ inline size_t SizeOf(DataType data_type) {
+  _(double, DataType::FLOAT64)        \
+  _(complex64, DataType::COMPLEX64)   \
+  _(complex128, DataType::COMPLEX128) \
+  _(pstring, DataType::STRING)
```
建议类型名和大写保持一致，pstring -> DataType::PSTRING，或者string -> DataType::STRING

joey12300 replied:
已修改~

joey12300 committed:
STRING->PSTRING; Add CreateInferLikeMeta

joey12300 committed:
Add code gen for C++ String API

joey12300 committed:
merge develop

joey12300 committed:
remove strings_api_utils.h

joey12300 committed:
Add ignore file (strings_api.h, strings_api.cc)

joey12300 committed:
update strings gen script

joey12300 committed:
merge develop

joey12300 committed:
change args order of case convert kernels

joey12300 committed:
Add comments for pstring, StringTensor

joey12300 committed:
cpstring_internal.h -> cpstring_impl.h

chenwhql commented on paddle/phi/kernels/strings/gpu/case_convert_kernel.cu:
```diff
@@ -0,0 +1,182 @@
+/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
license格式好像有点问题，空行缩进可以调整下

joey12300 replied:
已修改。

chenwhql commented on paddle/phi/kernels/strings/gpu/strings_copy_kernel.cu:
```diff
@@ -0,0 +1,150 @@
+  const auto& src_place = src.place();
+  auto dst_place = dst->place();
+
+  if (src_place == dst_place && paddle::platform::is_cpu_place(src_place)) {
```
这里不建议在phi下继续使用platform下接口，可以直接使用`place.GetType() == phi::AllocationType::GPU`判断

joey12300 replied:
所有涉及place判断的地方都用`place.GetType() == phi::AllocationType::xxx`替代了。

chenwhql commented on paddle/phi/kernels/strings/gpu/strings_copy_kernel.cu:
```diff
@@ -0,0 +1,150 @@
+  auto dst_place = dst->place();
+
+  if (src_place == dst_place && paddle::platform::is_cpu_place(src_place)) {
+    PADDLE_THROW(paddle::platform::errors::InvalidArgument(
```
paddle::platform::errors -> phi::errors，其他有的话也替换下

joey12300 replied:
已将所有`paddle::platform`替换为`phi::`

chenwhql commented on paddle/phi/kernels/strings/gpu/strings_copy_kernel.cu:
```diff
@@ -0,0 +1,150 @@
+  if (paddle::platform::is_gpu_place(src_place) &&  // NOLINT
+      paddle::platform::is_cpu_place(dst_place)) {
+    // Situation 1: gpu_place->cpu_place
+    phi::DeviceContextPool& pool = phi::DeviceContextPool::Instance();
```
phi kernel中不可以使用单例，否则推理无法使用，这种写法需要升级，应该可以规避，参考DenseTensor的copy_kernel

joey12300 replied:
已修改，将单例从kernel中删除

chenwhql commented on paddle/phi/kernels/strings/strings_copy_kernel.h:
```diff
@@ -0,0 +1,32 @@
+#include "paddle/phi/core/string_tensor.h"
+
+// See Note [ Why still include the fluid headers? ]
+#include "paddle/fluid/memory/memcpy.h"
```
放到需要的.cc/cu中，头文件不需要

joey12300 replied:
已删除所有fluid头文件的引用（当时希望使用is_gpu_place等接口，所以引入fluid头文件）

chenwhql commented on paddle/phi/kernels/strings/unicode.cc:
```diff
@@ -0,0 +1,89 @@
+static const void* utils_map[4] = {nullptr};
+static uint16_t CHARCASES_MAP[65536] = {0};
+
+const uint8_t* get_uniflag_map() {
```
保持code style一致，函数采用驼峰式命名？

joey12300 replied:
已使用Camel Code Style修改这些函数的命名

joey12300 committed:
Update accordding to comments:

1. Remove fluid headers
2. paddle::platform::errors -> phi::errors
3. Use 'place.GetType() == phi::AllocationType::GPU' instead of 'paddle::platform::is_cpu_space()'
4. Use camel code style

joey12300 committed:
Remove all singletons in strings kernels

Shixiaowei02 commented:
Recommended things to do
1. Add high-level APIs for inference.
2. Test on mobile.

joey12300 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_string_tensor

wawltor commented on paddle/fluid/framework/convert_utils.cc:
```diff
@@ -81,6 +81,8 @@ paddle::framework::proto::VarType::Type TransToProtoVarType(
     case DataType::BOOL:
       return paddle::framework::proto::VarType::BOOL;
+    case DataType::PSTRING:
+      return paddle::framework::proto::VarType::RAW;
```
这块为什么返回RAW类型的数据了？

joey12300 replied:
已添加PSTRING  prototype，改为返回`paddle::framework::proto::VarType::RAW`

wawltor commented on paddle/phi/api/lib/api_gen_utils.cc:
```diff
@@ -56,6 +56,10 @@ std::shared_ptr<phi::SelectedRows> TensorToSelectedRows(
+std::shared_ptr<phi::StringTensor> TensorToStringTensor(const Tensor& tensor) {
+  return std::dynamic_pointer_cast<phi::StringTensor>(tensor.impl());
+}
+
```
没有太get这个函数的主要功能是什么？

wawltor commented on paddle/phi/common/cpstring_impl.h:
```diff
@@ -0,0 +1,541 @@
+/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
这里的Copyright建议加上Tensorflow相关CopyRight

joey12300 replied:
已添加

wawltor commented on paddle/phi/common/layout.h:
```diff
@@ -29,6 +29,7 @@ enum class DataLayout {
   MKLDNN,
   SPARSE_COO,
   SPARSE_CSR,
+  PSTRING,
```
如果PSTRING作为一个基础的数据类型，为什么这里还要作为一个LayOut

joey12300 replied:
已将 `DataLayout::PSTRING`改为 `DataLayout::PSTRING_UNION`

wawltor commented on paddle/phi/core/tensor_meta.h:
```diff
@@ -64,4 +64,25 @@ inline bool operator==(const DenseTensorMeta& lhs, const DenseTensorMeta& rhs) {
+
+inline bool operator==(const StringTensorMeta& lhs,
+                       const StringTensorMeta& rhs) {
+  bool ret = true;
```
ret看起来这个变量是多余的

joey12300 replied:
已删除

wawltor commented on paddle/phi/core/tensor_meta.h:
```diff
@@ -64,4 +64,25 @@ inline bool operator==(const DenseTensorMeta& lhs, const DenseTensorMeta& rhs) {
+inline bool operator==(const StringTensorMeta& lhs,
+                       const StringTensorMeta& rhs) {
+  bool ret = true;
+  return ret && (lhs.is_scalar == rhs.is_scalar) && (lhs.dims == rhs.dims) &&
```
这里有个为啥需要与的关系，必须是scalar吗？

joey12300 replied:
该函数判断`StringTensorMeta `所有成员变量是否一致，所以用与。

wawltor commented on paddle/phi/core/string_tensor.cc:
```diff
@@ -0,0 +1,156 @@
+
+  if (!holder_ || holder_->size() < bytes + meta_.offset) {
+    meta_.offset = 0;
+    VLOG(10) << "Allocate data with bytes: " << bytes;
```
这里的日志打上string tensor相关的信息，便于debug

joey12300 replied:
已添加。

wawltor commented on paddle/phi/kernels/strings/cpu/case_convert_kernel.cc:
```diff
@@ -0,0 +1,56 @@
+/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
为啥文件名是case开头了？

joey12300 replied:
已使用strings_lower_upper替换case_convert_

wawltor commented on paddle/phi/kernels/strings/case_convert_kernel.h:
```diff
@@ -0,0 +1,122 @@
+/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
整体检查一下文件名、函数名，case开头有点奇怪

wawltor commented on paddle/phi/kernels/strings/cpu/strings_copy_kernel.cc:
```diff
@@ -0,0 +1,57 @@
+          StringTensor* dst) {
+  auto* src_ptr = src.data();
+  const auto& src_place = src.place();
+
```
这里的日志可以打上String tensor的相关flag，便于debug

joey12300 replied:
已添加String tensor copy相关信息。

wawltor commented on paddle/phi/kernels/strings/cpu/strings_serialize_kernel.cc:
```diff
@@ -0,0 +1,35 @@
+void Serialize(const Context& dev_ctx,
+               const StringTensor& src,
+               DenseTensor* dst) {
+  SerializeCPUKernel(dev_ctx, src, dst);
```
为什么需要把序列化做成一个op kernel了，序列化部分在dump模型不是以kernel方式来dump，而不是functional的方式

joey12300 replied:
已将Serialize，Deserialize从kernel转成function的形式。

wawltor commented on paddle/phi/kernels/strings/gpu/case_convert_kernel.cu:
```diff
@@ -0,0 +1,180 @@
+    uint32_t* unicode_output_ptr =
+        thrust::raw_pointer_cast(unicode_output.data());
+
+    CharConverter<GPUContext> converter(unicode_flag_map, cases_map);
```
这里将cpu的map拷贝到gpu，gpu上能直接访问一个hash_map吗？

joey12300 replied:
这里命名虽然是map，但数据结构是个数组。是一个unciode码与一些标记位的映射表

joey12300 committed:
fix rocm compile

joey12300 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_string_tensor

joey12300 committed:
Fix py3 compile

joey12300 committed:
Fix c++ coverage

joey12300 committed:
1. Add pstring proto type
2. Add StringTensor debug info
3. Rename case_convert_kernel to strings_lower_upper
4. Remove serialize derialize strings kernel

joey12300 committed:
DataLayout::PSTRING -> DataLayout::PSTRING_UNION

joey12300 committed:
Register pstring data type

joey12300 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_string_tensor

joey12300 committed:
Fix strings api gen

wawltor reviewed (dismissed):
LGTM

chenwhql commented on paddle/phi/tests/core/test_string_tensor.cc:
```diff
@@ -0,0 +1,112 @@
+/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
细节，还有几个文件的license没有换行

chenwhql reviewed (dismissed):
LGTM

joey12300 committed:
merge

joey12300 committed:
Fix dense tensor register pstring dtype

joey12300 committed:
Fix error messages

XiaoguangHu01 reviewed (dismissed):
LGTM

joey12300 committed:
merge

joey12300 committed:
remove line

joey12300 committed:
add pstring unittest

joey12300 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_string_tensor

joey12300 committed:
remove test string api unitest

joey12300 committed:
remove empty line

joey12300 committed:
Remove some headers to decrease the size of executable file

ZeyuChen reviewed (approved):
LGTM
